bbl_screen-patch/patches/printerui/qml/x1plus/GcodeGenerator.js

```diff
@@ -490,62 +490,96 @@ const Tramming = {
-}
 
 /**
- * Print speed Gcode - M204.2 (acceleration magnitude), M220 (feed rate), and 
- * M73.2 (time remaining parameter)
+ * Print speed adjustment Gcode
+ * 
+ * M204.2 K - sets acceleration magnitude
+ * M220 K - sets feed rate
+ * M73.2 R - updates timeline and time estimate
  * 
- * Usage: Input a string ["Silent", "Normal", "Sport", "Ludicrous"] to generate
- * Gcode for OEM speed profiles. Input an integer between 30 and 180 to generate
- * Gcode with interpolated parameters.
+ * Usage: input an even integer between 30 and 180 to generate speed adjustment Gcode
+ * OEM speed profiles: Silent=50%, Normal=100%, Sport=124%, Ludicrous=166%
  */
+
 function printSpeed(inputSpeed) {
-    let config;
-    if (typeof inputSpeed === 'string') {
-        config = speed_presets[inputSpeed];
-        if (!config) {
-            return "";
-        }
-    } else if (typeof inputSpeed === 'number') {
-        if (inputSpeed < 30 || inputSpeed > 180) {
-            inputSpeed = 100;
+    const bambuParams = {
+        speed: [50, 100, 124, 166],
+        acceleration: [0.3, 1.0, 1.4, 1.6],
+        feedRate: [0.7, 1.0, 1.4, 2.0],
+        labels: ["Silent", "Normal", "Sport", "Ludicrous"]
+    };
+
+    function speedFraction(speedPercent) {
+        return Math.floor(10000 / speedPercent) / 100;
+    }
+
+    function accelerationMagnitude(speedPercent) {
+        const sf = speedFraction(speedPercent);
+        return Math.exp((sf - 1.0191) / -0.8139);
+    }
+
+    function feedRate(speedPercent) {
+        return 6.426e-5 * speedPercent * speedPercent - 2.484e-3 * speedPercent + 0.654;
+    }
+
+    function interpolate(x, xPoints, yPoints) {
+        if (x <= xPoints[0]) return yPoints[0];
+        if (x >= xPoints[xPoints.length - 1]) return yPoints[yPoints.length - 1];
+        
+        for (let i = 0; i < xPoints.length - 1; i++) {
+            if (xPoints[i] <= x && x < xPoints[i+1]) {
+                const t = (x - xPoints[i]) / (xPoints[i+1] - xPoints[i]);
+                return yPoints[i] * (1 - t) + yPoints[i+1] * t;
+            }
         }
-        var speedFraction = speed_interpolation.speed_fraction(inputSpeed);
-        var accelerationMagnitude = speed_interpolation.acceleration_magnitude(speedFraction);
-        var feedRate = speed_interpolation.feed_rate(inputSpeed);
-        var level = speed_interpolation.level(accelerationMagnitude);
-        config = {
-            speed_fraction: speedFraction,
-            accelerationMagnitude: accelerationMagnitude,
-            feedRate: feedRate,
-            level: level > 7 ? 7 : level
+    }
+
+    function scaleFn(func, dataPoints) {
+        return (x) => {
+            const speeds = dataPoints.map(point => point[0]);
+            const values = dataPoints.map(point => point[1]);
+            const scalingFactor = interpolate(x, speeds, values) / func(x);
+            return func(x) * scalingFactor;
```

Comment:
haha wait dividing by func(x) and then multiplying through by func(x) is just the same as multiplying by 1, right?  so the func(x) is not used here at all, and this *only* does the linear interpolation, right?

Reply:
The ratio is between linear interpolation of the "theoretical" value vs the linear interpolation of the "actual" value. 

This happened by accident, and I am now realizing I should have just used cubic spline interpolation because interpolating interpolated data is not exactly logical to me. 

This results in a smoothed linear interpolation very similar to cubic spline interpolation 
<img width="574" alt="image" src="https://github.com/user-attachments/assets/dbfb7e59-7e58-402e-aa9f-2a945466f657">